### PR TITLE
fix: bg audio on silence labels at eval/test set

### PR DESCRIPTION
**Pull request by stocyr**
The model is never trained on pure zeros input, so depending on the convolution biases, the model consistently gets the __silence__ label wrong during clean eval/test set evaluation.
So i recommend either occasionally training with all-zeros for the __silence__ label or (what I did in the PR) only evaluate on at least *some* background noise for the __silence__ label.

This has also been changed recently in the original repo with tensorflow/tensorflow@024b037

### CODE REVIEW

**Comment by daemon** [comment]
Thanks for the fix. I had mirrored the training and validation process as close as possible to the TF implementation. We didn't observe wrong labels for silence while testing our models -- was this on `cnn-trad-fpool2`? What final model accuracy do you get after your changes?

**Comment by daemon** [comment]
I noticed that this change uses the same background noise for both training and dev/test. It would be a good idea to partition the background noise audio into training/dev/test sets, e.g., first 80% of `washing_dishes.wav` is for training, next 10% for dev, and final 10% for test.

**Comment by stocyr** [comment]
Partitioning the background noise wouldn't be a bad idea, indeed. However, it's not that crucial since we don't have discrete noise examples that would look exactly the same - just random temporally shifted ones.

Regarding the model accuracy: I had one where the training acc was around 91% and the test/dev acc at 83% (pretty much 1/12, so one "complete" label). The confusion matrix looked suspicious enough to delve into the problem:
confmat_615n_sweep_mel_test

**Comment by daemon** [comment]
Cool, thanks for sharing. This is on a fully converged `cnn-trad-fpool2` model, right? Our [reproduction](https://arxiv.org/pdf/1710.06554.pdf) didn't seem to be affected at the time, for whatever reason.

**Comment by stocyr** [comment]
(Oh, sorry for the four-fold plot, only the first one matters)
The model is a fully converged `cnn-trad-pool2` (best validation score after 46 epochs, test score 89.8).
If I'd hoover over that wrong placed silence pixel in the confusion matrix, that's a full 100%; if the model just gets fed a pure zero'ed out image, it just extracts the bias from the first conv layer and feeds that through the second layer, so we're just stuck with a fixed model state for 1/12 of all examples! And if it never _had_ to learn the mapping of a zero-image to the `silence` label, it has no direct reason to do so during eval/test.

But as you say, and I think this follows intuition, the randomness during training can cause a model to just "get the biases" right for the first conv layer, and then you're good.
I then recently discovered that the tensorflow implementation fixed _just that_ in April 12th...

**Comment by daemon** [comment]
Ah, I see. What results do you get after the changes?
